### ISSUE TITLE
Add "is_reachable" to host object

### ISSUE THREAD

**Issue by thomseddon** [Icinga/icinga2#6844]
## Expected Behavior
When requesting a host object via the api (https://x.x.x.x:5665/v1/objects/hosts?hosts=1234), the `is_reachable` field should be returned

## Current Behavior
The `is_reachable` field is not returned. 

## Possible Solution
Include `is_reachable` in API response.

## Steps to Reproduce (for bugs)

1. `curl -k -s -X GET -H 'Accept: application/json' -H 'Content-Type: application/json' -u xx:xx 'https://x.x.x.x:5665/v1/objects/hosts?hosts=1234'`
2.
3.
4.

## Context
For example, the host below is not reachable due to a failed dependency.
We have "disable_checks" enabled and as the dependency failed after the last check, no further checks were executed. As the last executed check succeeded there is no way to determine that the host is unreachable via the API.

icingaweb2 shows the host is not reachable as determined by the "is_reachable" field which is available via the IDO interface but not the API :(

```
# curl -k -s -X GET -H 'Accept: application/json' -H 'Content-Type: application/json' -u xx:xx 'https://x.x.x.x:5665/v1/objects/hosts?hosts=1988' | jq
{
  "results": [
    {
      "attrs": {
        "__name": "1988",
        "acknowledgement": 0,
        "acknowledgement_expiry": 0,
        "action_url": "",
        "active": true,
        "address": "10.0.0.1",
        "address6": "",
        "check_attempt": 1,
        "check_command": "ping4",
        "check_interval": 30,
        "check_period": "",
        "check_timeout": null,
        "command_endpoint": "",
        "display_name": "hostname123",
        "downtime_depth": 0,
        "enable_active_checks": true,
        "enable_event_handler": true,
        "enable_flapping": true,
        "enable_notifications": true,
        "enable_passive_checks": true,
        "enable_perfdata": true,
        "event_command": "",
        "flapping": false,
        "flapping_current": 0,
        "flapping_last_change": 1515772203.052052,
        "flapping_threshold": 0,
        "flapping_threshold_high": 30,
        "flapping_threshold_low": 25,
        "force_next_check": false,
        "force_next_notification": false,
        "groups": [
          "hosts-priority-high"
        ],
        "ha_mode": 0,
        "icon_image": "",
        "icon_image_alt": "",
        "last_check": 1545096024.779654,
        "last_check_result": {
          "active": true,
          "check_source": "example.telcom.network",
          "command": [
            "/usr/lib64/nagios/plugins/check_ping",
            "-4",
            "-H",
            "10.0.0.1",
            "-c",
            "5000,100%",
            "-p",
            "5",
            "-w",
            "500,100%"
          ],
          "execution_end": 1545096024.779608,
          "execution_start": 1545096020.74619,
          "exit_status": 0,
          "output": "PING OK - Packet loss = 0%, RTA = 19.74 ms",
          "performance_data": [
            "rta=19.742001ms;500.000000;5000.000000;0.000000",
            "pl=0%;100;100;0"
          ],
          "schedule_end": 1545096024.779654,
          "schedule_start": 1545096020.745869,
          "state": 0,
          "type": "CheckResult",
          "vars_after": {
            "attempt": 1,
            "reachable": true,
            "state": 0,
            "state_type": 1
          },
          "vars_before": {
            "attempt": 1,
            "reachable": true,
            "state": 0,
            "state_type": 1
          }
        },
        "last_hard_state": 0,
        "last_hard_state_change": 1544816401.983204,
        "last_reachable": true,
        "last_state": 0,
        "last_state_change": 1544816401.983204,
        "last_state_down": 1544816260.430792,
        "last_state_type": 1,
        "last_state_unreachable": 1543767570.038673,
        "last_state_up": 1545096024.779696,
        "max_check_attempts": 3,
        "name": "1988",
        "next_check": 1545147302.29,
        "notes": "",
        "notes_url": "",
        "original_attributes": null,
        "package": "_etc",
        "paused": false,
        "retry_interval": 10,
        "severity": 8,
        "source_location": {
          "first_column": 1,
          "first_line": 7081,
          "last_column": 18,
          "last_line": 7081,
          "path": "/etc/icinga2/zones.d/masters/hosts.conf"
        },
        "state": 0,
        "state_type": 1,
        "templates": [
          "1988",
          "host-priority-high"
        ],
        "type": "Host",
        "vars": {
          "location_id": "8",
          "location_name": "A Location, UK",
          "ping_cpl": 100,
          "ping_crta": 5000,
          "ping_packets": 5,
          "ping_wpl": 100,
          "ping_wrta": 500,
          "priority": 2,
          "priority_str": "example",
          "sla": "24x7",
          "subnet_id": "310"
        },
        "version": 0,
        "volatile": false,
        "zone": "masters"
      },
      "joins": {},
      "meta": {},
      "name": "1988",
      "type": "Host"
    }
  ]
}
```

## Your Environment
* Version used (`icinga2 --version`): r2.8.1-1
* Operating System and version: Amazon Linux AMI
* Enabled features (`icinga2 feature list`): api checker command ido-mysql mainlog notification
* Icinga Web 2 version and modules (System - About): 2.4.1
* Config validation (`icinga2 daemon -C`): N/A
* If you run multiple Icinga 2 instances, the `zones.conf` file (or `icinga2 object list --type Endpoint` and `icinga2 object list --type Zone`) from all affected nodes. N/A

**Comment by dnsmichi** [comment]
`last_reachable` is the attribute you're looking for.

**Comment by dnsmichi** [comment]
Reachability calculation is pretty expensive as it involves checking the dependency chain recursively. One needs to use a cached value here, which is available with `last_reachable` being updated when processing a new CheckResult. If this would be made a direct query attribute in `is_reachable`, the API would be slowed down. 

In terms of Icinga Web, the data is stored only once in the database where the web reads from. This also happens when a new checkresult generates an update to the *status tables in the database.

**Comment by thomseddon** [comment]
In the above examples, checks are executed every 30 seconds but `last_reachable` is still `true` over 24 hours later.

So, it seems to me that this value is not being correctly updated if the device is unreachable and `disable_checks` is set to `true`.

I would propose: **If a host or service a unreachable, and `disable_checks` is true, `last_reachable` is still updated to be false.** ?

**Comment by lippserd** [comment]
We'll have a look why last_reachable (API) differs from is_reachable (IDO) since both should be the same. But it will take us some time.

**Comment by thomseddon** [comment]
Brilliant, thank you! I'll keep an eye on this issue.

**Comment by dnsmichi** [comment]
DB IDO uses `IsReachable()` while it should use `GetLastReachable()` instead.

**Comment by dnsmichi** [comment]
hostdbobject calls this twice, servicedbobject once. In scenarios with many objects and dependency trees this likely degrades performance on status updates.

**Comment by thomseddon** [comment]
@dnsmichi I believe #7163 means that the DB will now show the same behaviour as the API, will there be a fix to ensure the API (and now the DB) will return the correct value, my example above shows that the value was incorrect?

I believe the value is incorrect because `disable_checks = true` is set, and so the reachability is not being updated. If this is correct, then I believe this is bug?

**Comment by dnsmichi** [comment]
My fix aligns the behaviour of both interfaces. Our conclusion to this issue (and others) was that calculating the dependency tree on each event is far too expensive, doesn't matter whether this happens via HTTP client or IDO receiver. The cached reachability calculated on processing a check result is the correct metric here, there's no "live" view available.

**Comment by thomseddon** [comment]
I agree the cached value is preferable, but in my example the cached value is clearly incorrect.

I think there are two options:

- Ensure the cached value is correct
- Add documentation explaining when/why this value will be incorrect

In terms of a fix, my theory is that it's not updated to be unreachable as disable_checks is true - do you agree?

**Comment by dnsmichi** [comment]
I'll add a note to the docs, I have some changes for 2.11 in my stash anyways. Agreed, disable_checks is not a good choice here then. Imho that attribute is a design failure since checks are wanted quite often - but no notifications.

**Comment by thomseddon** [comment]
Okie dokie, thanks. On the design of disable_checks, I somewhat agree.

Anecdotally, the reason we use disable_checks is because without this, we get a lot of notification noise from dependants after a parent recovers.

For example, with the following dependencies A <- B <- C, if A goes down, we will get one notification for A. However, when it recovers we will get notifications for B and then C (I assume this is because notifications are no longer suppressed but the last check performed, during the dependency failure, was DOWN).
Equally, we have escalations after a certain amount of time, with the above issue escalations would be triggered immediately once the parent recovers, then resolved following the first successful check - with larger incidents this creates a LOT of pager noise :)

Should I open this as a separate bug or is this a known issue/limitation?
